ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,171 @@
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_watershed_streams(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    # `watershed_lonlats` must be an ordered collection (not sets) because
+    # a multi line string is an array (python list) of linestrings
+    watershed_lonlats = [[]]
+    counter = 0
+
+    for frozen in watershed_xys:
+        for xy in frozen:
+            watershed_lonlats[counter].append(flow_direction.xy_to_lonlat(xy))
+        counter += 1
+        watershed_lonlats.append([])
+
+    if watershed_lonlats[-1] == []:
+        watershed_lonlats.remove([])
+
+    lines = MultiLineString(watershed_lonlats)
+
+    return {
+        "streams": geojson_feature(lines,),
+        "debug/test": {"watershed_streams": {"time": watershed_time.elapsed,}},
+    }
+
+
+def build_watershed_streams(target, routing, direction_map, debug=False):
+    """
+    Return set of all paths of cells that drain into `target`.
+
+    :param target: An xy index representing the cell of interest.
+    :param routing: A numpy array representing water flow using VIC direction
+        codes (0 - 9). These codes may be integers or floats.
+    :param direction_map: Maps VIC direction codes into offsets in cell indices.
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Set of frozensets of cells (streams) that drain into `target`.
+
+    Notes:
+
+    - In this function, a cell is represented by an (x, y) index pair.
+
+    - Routing graphs can and in practice do contain cycles. Variable `visited`
+    is used to determine whether a cell has already been visited during the
+    traversal of the routing graph, i.e., whether we are cycling, and if so
+    not to repeat that subgraph.
+    """
+    visited = set()
+    i = 0
+    connection = [[]]
+
+    def upstream(cell):
+        """Return graph description of how the stream is connected
+        by movement of water"""
+        nonlocal visited
+        nonlocal i
+        nonlocal connection
+
+        visited |= {cell}
+        connection[i].append(cell)
+        for neighbour in neighbours(cell):
+            if neighbour not in visited and is_upstream(
+                neighbour, cell, routing, direction_map
+            ):
+                upstream(neighbour)
+                connection.append([])
+                i += 1
+                connection[i].append(cell)
+        return connection
+
+    connection = upstream(target)
+    return set(frozenset(i) for i in connection if len(i) > 1)
```

Comment:
@helfy18 , kindly have a look at the [discussion in the issue](https://github.com/pacificclimate/climate-explorer-backend/issues/198#issuecomment-1054706359) regarding what linestrings should be returned. I think this code solves a more difficult and unnecessary problem. Could you revise it to solve the simpler problem?